push/sender/src/main/java/org/jboss/aerogear/unifiedpush/message/sender/GCMPushNotificationSender.java

```diff
@@ -105,30 +113,57 @@ public void sendPushMessage(Variant variant, Collection<String> tokens, UnifiedP
             final Sender sender = new Sender(androidVariant.getGoogleKey());
 
             // send out a message to a batch of devices...
-            processGCM(androidVariant, registrationIDs, gcmMessage, sender);
-
+            if (registrationIDs.size() > 0) {
+                processDirectMessages(androidVariant, registrationIDs, gcmMessage, sender);
+            }
+            
+            if (topics.size() > 0) {
+                processTopicMessages(androidVariant, topics, gcmMessage, sender);
+            }
+            
             logger.info("Message batch to GCM has been submitted");
             callback.onSuccess();
 
         } catch (Exception e) {
             // GCM exceptions:
-            logger.severe("Error sending payload to GCM server");
+            logger.severe("Error sending payload to GCM server", e);
             callback.onError("Error sending payload to GCM server");
         }
     }
 
     /**
-     * Process the HTTP POST to the GCM infrastructor for the given list of registrationIDs.     *
+     * Process the HTTP POST to the GCM infrastructure for the given list of registrationIDs.     *
      */
-    private void processGCM(AndroidVariant androidVariant, List<String> registrationIDs, Message gcmMessage, Sender sender) throws IOException {
+    private void processDirectMessages(AndroidVariant androidVariant, List<String> registrationIDs, Message gcmMessage, Sender sender) throws IOException {
 
         logger.info("Sending payload for [" + registrationIDs.size() + "] devices to GCM");
 
-        MulticastResult multicastResult = sender.send(gcmMessage, registrationIDs, 0);
+        MulticastResult multicastResult = sender.sendNoRetry(gcmMessage, registrationIDs);
 
         // after sending, let's identify the inactive/invalid registrationIDs and trigger their deletion:
         cleanupInvalidRegistrationIDsForVariant(androidVariant.getVariantID(), multicastResult, registrationIDs);
     }
+    
+    /**
+     * Process the HTTP POST to the GCM infrastructure for the given list of topics.
+     */
+    private void processTopicMessages(AndroidVariant androidVariant, List<String> topics, Message gcmMessage, Sender sender) throws IOException {
+
+        logger.info("Sending payload for [" + topics.size() + "] topics to GCM");
+
+        for (String topic : topics) {
+            Result result = sender.sendNoRetry(gcmMessage, topic);
+            if (Constants.ERROR_TOPIC_TOO_MANY_SUBSCRIBERS.equals(result.getErrorCodeName())) {
+                logger.severe("Topic '" + topic + "' has more than one million subscribers and must be resent as individual messages.");
+                //TODO: Resend as individual messages
```

Comment:
@lfryc I know you worked on most of the JMS stuff.  What is the best way to signal back to UPS that we need to run this task as individual messages instead of a topic message? (Let's assume I can do all the GCM data aggregation and just need to know how to send the message and what metadata would be useful)